### PR TITLE
[ViT_MAE] fix num of channels in `patchify` and `unpatchify`

### LINKED ISSUE #17473: Number of channels in the ViTMAE model
### Feature request

Dear huggingface community,
I am experimenting with the ViTMAE model from the transformers library. The ViTMAEConfig class has the option "num_channels" to specify the number of input (color) channels belonging to an image. If I modify this, say, to 1 (for processing grayscale images), the model throws an error, due to the number "3" being hard-coded into the functions "patchify" and "unpatchify" in the file "modeling_vit_mae.py". 

### Motivation

I would like to request to change this such that any number of input channels is possible. 

### Your contribution

As noted above, one only has to change the functions "patchify" and "depatchify" to either scan for the input color channels, or provide the input color channels as a class variable such that both functions can use it (instead of the hard-coded value "3"). I checked this and on my system it worked out just fine.

### PR DESCRIPTION
**Pull request by johnnv1**
# What does this PR do?
Fix the hard-coded number of channels in `patchify` and `unpatchify` methods.

Fixes #17473 

## Who can review?
@NielsRogge

### CODE REVIEW

File: src/transformers/models/vit_mae/modeling_vit_mae.py [hunk 1]
```diff
@@ -870,33 +870,33 @@ class PreTrainedModel
 
     def patchify(self, imgs):
         """
-        imgs: (N, 3, H, W) x: (N, L, patch_size**2 *3)
+        imgs: (N, num_channels, H, W) x: (N, L, patch_size**2 * num_channels)
```
Comment (NielsRogge) [inline src/transformers/models/vit_mae/modeling_vit_mae.py:873]:
```suggestion
        imgs: (batch_size, num_channels, height, width)
```

File: src/transformers/models/vit_mae/modeling_vit_mae.py [hunk 2]
```diff
@@ -870,33 +870,33 @@ class PreTrainedModel
...
+        x = imgs.reshape(shape=(imgs.shape[0], self.config.num_channels, h, p, w, p))
         x = torch.einsum("nchpwq->nhwpqc", x)
-        x = x.reshape(shape=(imgs.shape[0], h * w, p**2 * 3))
+        x = x.reshape(shape=(imgs.shape[0], h * w, p**2 * self.config.num_channels))
         return x
 
     def unpatchify(self, x):
         """
-        x: (N, L, patch_size**2 *3) imgs: (N, 3, H, W)
+        x: (N, L, patch_size**2 * num_channels) imgs: (N, num_channels, H, W)
```
Comment (NielsRogge) [inline src/transformers/models/vit_mae/modeling_vit_mae.py:886]:
```suggestion
        x: (batch_size, seq_length, patch_size**2 * num_channels)
```

File: src/transformers/models/vit_mae/modeling_tf_vit_mae.py [hunk 3]
```diff
@@ -971,39 +971,41 @@ def _prune_heads(self, heads_to_prune):
...
-        imgs = tf.reshape(x, (shape_list(x)[0], h * p, h * p, 3))
+        imgs = tf.reshape(x, (shape_list(x)[0], h * p, h * p, self.config.num_channels))
         return imgs
 
     def forward_loss(self, imgs, pred, mask):
         """
-        imgs: [batch_size, height, width, 3] pred: [batch_size, num_patches, patch_size**2*3] mask: [N, L], 0 is keep,
-        1 is remove,
+        imgs: [batch_size, height, width, num_channels] pred: [batch_size, num_patches, patch_size**2 * num_channels]
+        mask: [N, L], 0 is keep, 1 is remove,
```
Comment (NielsRogge) [inline src/transformers/models/vit_mae/modeling_tf_vit_mae.py:1008]: Can you use the docs style guide explained here: https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation

File: src/transformers/models/vit_mae/modeling_vit_mae.py (hunk 2 above)
Comment (NielsRogge) [inline src/transformers/models/vit_mae/modeling_vit_mae.py:886]: Same comment here

File: src/transformers/models/vit_mae/modeling_vit_mae.py (hunk 1 above)
Comment (NielsRogge) [inline src/transformers/models/vit_mae/modeling_vit_mae.py:873]: Same comment here

**Comment by johnnv1** [comment]
@NielsRogge I didn't find what is breaking the doc now

**Comment by johnnv1** [comment]
@NielsRogge , can you re-run the gh action? looks like something interrupted it

**Comment by NielsRogge** [comment]
Hi,

I've created PR #17710 that fixes some more things, like the variable names and docstrings.

**Pull request closed without merging**